orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/CompoundExecutionOperator.java

```diff
@@ -79,32 +78,56 @@ public void resume(
       @Nullable String user,
       @NonNull Boolean ignoreCurrentStatus) {
     doInternal(
-        (PipelineExecution execution) -> runner.unpause(execution),
+        runner::unpause,
         () -> repository.resume(executionType, executionId, user, ignoreCurrentStatus),
         "resume",
         executionType,
         executionId);
   }
 
-  private void doInternal(
+  public PipelineExecution updateStage(
+      @NonNull ExecutionType executionType,
+      @NonNull String executionId,
+      @NonNull String stageId,
+      @NonNull Consumer<StageExecution> stageUpdater) {
+    return doInternal(
+        runner::reschedule,
+        () -> {
+          PipelineExecution execution = repository.retrieve(executionType, executionId);
+          StageExecution stage = execution.stageById(stageId);
+
+          // mutates stage in place
+          stageUpdater.accept(stage);
+
+          repository.storeStage(stage);
+        },
+        "reschedule",
+        executionType,
+        executionId);
+  }
+
+  private PipelineExecution doInternal(
```

Comment:
wait.. this just occurred to me... doesn't this method pose a race?
we push a message onto the queue before we mutate the repository. so if the message gets picked up before we manage to save the repo we are going to be in a weird state where the stage/execution have the right data but the action didn't take place because the queue processed the message before the state was updated?

i think we can just reverse the order and we will be fine

Reply:
You're right, fixed and tested in 0cfef4d0d!